### PR TITLE
"implemented test features and author sort"

### PR DESCRIPTION
**Pull request by jhcoll**

### CODE REVIEW

File: src/test/java/com/scottlogic/AuthorPostSorterTest.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+import java.time.ZoneOffset;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class AuthorPostSorterTest {
+    @Test
+    void nullTest() {
```
Comment (jluxford-scottlogic) [inline src/test/java/com/scottlogic/AuthorPostSorterTest.java]: As per the talk with Matt a good style for test names is `function_inputValues_expectedOutput` as this makes it clearer when it flags as a failure when running the full test suite.

File: src/test/java/com/scottlogic/AuthorPostSorterTest.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+        assertNull(actual);
+    }
+
+    @Test
+    void emptyListTest() {
+        List<UserPost> actual = Collections.emptyList();
+        List<UserPost> expected = Collections.emptyList();
+
+        actual = new AuthorPostSorter().sort(actual);
```
Comment (jluxford-scottlogic) [inline src/test/java/com/scottlogic/AuthorPostSorterTest.java]: In general you don't want side effects and mutation inside tests. Instead of setting `actual` to be `Collections.emptyList()` and then also the sorted version having a variable that is `initialList` or something similar makes the tests significantly clearer. The style of testing is different to normal coding as it is all about maximizing readability rather than efficiency as they will be run a finite number of times, this means a slight efficiency lose doesn't have much impact.

File: src/test/java/com/scottlogic/AuthorPostSorterTest.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+        assertEquals(expected, actual);
+    }
+
```
Comment (jluxford-scottlogic) [inline src/test/java/com/scottlogic/AuthorPostSorterTest.java]: There are a few easier test cases that you might want to do before moving onto 3 length lists

File: src/test/java/com/scottlogic/AuthorPostSorterTest.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        UserPost userPost3 = new UserPost("jane smith",
+                OffsetDateTime.of(2020, 3, 12, 13, 22, 12, 0, ZoneOffset.UTC),
+                "An example of a post \nwith lines breaks.", 3);
+
+        List<UserPost> actual = Arrays.asList(userPost1, userPost2, userPost3);
+        List<UserPost> expected = Arrays.asList(userPost3, userPost1, userPost2);
+
+        actual = new AuthorPostSorter().sort(actual);
+
+        assertEquals(expected, actual);
```
Comment (jluxford-scottlogic) [inline src/test/java/com/scottlogic/AuthorPostSorterTest.java]: Nice coverage of the upper and lower case but you can also try matching them

File: src/test/java/com/scottlogic/AuthorPostSorterTest.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    }
```
Comment (jluxford-scottlogic) [inline src/test/java/com/scottlogic/AuthorPostSorterTest.java]: There might be a few fringe and edge cases that might be worth investigating, if you have a go at trying to find a few of them and if you run out I can see what other cases might be useful.

**Review by jluxford-scottlogic — APPROVED** [review]
Nice bit of code, a good start with the testing, there are few style and general changes that can be done and a few cases that might be useful to add

File: src/main/java/com/scottlogic/AuthorPostSorter.java [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+
+public class AuthorPostSorter implements PostSorter {
+
+    @Override
+    public List<UserPost> sort(List<UserPost> inputList) {
+        if(inputList != null){
+                inputList.sort(Comparator.comparing(UserPost::getAuthor));
```
Comment (jluxford-scottlogic) [inline src/main/java/com/scottlogic/AuthorPostSorter.java]: I missed this on first pass, but this is mutating the `inputList` before returning it. In general you shouldn't mutate input values as when this causes bugs they are really hard to debug

File: src/test/java/com/scottlogic/AuthorPostSorterTest.java (hunk 1 above)
Comment (jhcoll) [reply src/test/java/com/scottlogic/AuthorPostSorterTest.java]: Yeah makes sense

File: src/test/java/com/scottlogic/AuthorPostSorterTest.java (hunk 2 above)
Comment (jhcoll) [reply src/test/java/com/scottlogic/AuthorPostSorterTest.java]: This makes sense, and avoids accidentally incorrectly manipulating the actual

File: src/main/java/com/scottlogic/AuthorPostSorter.java (hunk 6 above)
Comment (jhcoll) [reply src/main/java/com/scottlogic/AuthorPostSorter.java]: Yeah will be sure to change that
